### DISCUSSION TITLE
Resampling unobserved variables in `sample_posterior_predictive`

### DISCUSSION THREAD

**Discussion by zaxtax** [pymc-devs/pymc3#7225] [Ideas]
As part of using `pmx.as_model`, I often end up writing code that looks a like this like:

````python
import pymc as pm
import pymc_extras as pmx
import numpy as np

n_obs = 100
true_beta = 2.5
true_alpha = 0.25

x = np.random.normal(size=n_obs)
true_p = 1 / (1 + np.exp(-(true_alpha + true_beta * x)))
y = np.random.binomial(n=1, p=true_p)

@pmx.as_model()
def logistic_model(x, y):
    alpha = pm.Normal("alpha")
    beta = pm.Normal("beta")
    p = pm.Deterministic("p", pm.math.sigmoid(alpha + beta * x))
    obs = pm.Bernoulli("obs", p=p, observed=y, shape=x.shape[0])

lm = logistic_model(x, y)
idata = pm.sample(model=lm)

grid_size = 250
x_oos = np.linspace(x.min(), x.max(), grid_size)
lm_oos = logistic_model(x_oos, None)
oos_idata = pm.sample_posterior_predictive(
    idata,
    model=lm_oos,
    var_names=["p", "obs"],
    predictions=True, 
)
````

This is fairly clean. I don't have to put a dummy value for `y` when doing oos. But, I have to pass both `p` and `obs` to `var_names`. Setting aside the need to pass `p` to force resampling of the Deterministic. I don't know why we need to specify `obs`. While it's not in `lm_oos` it is in `idata` and it would be nice to resample it!

What do others thinks?

**Comment by twiecki** [comment]
I agree, that would be nice.

For now, you could just not specify `observerd` and only add it with `pm.observe()` at inference time and otherwise have the first model just be the generative one.

```python
@pmx.as_model()
def logistic_model(x):
    alpha = pm.Normal("alpha")
    beta = pm.Normal("beta")
    p = pm.Deterministic("p", pm.math.sigmoid(alpha + beta * x))
    obs = pm.Bernoulli("obs", p=p) # I don't think you even need shape here

lm = logistic_model(x)
lm_obs = pm.observe(lm, {"obs": y})
idata = pm.sample(model=lm_obs)
```

**Comment by zaxtax** [comment]
But will I need to add it while generating predictions?

**Reply by twiecki** [reply]
Oh, I think I misunderstood. Are you saying you want `sample_posterior_predictive` to look for what variable is observed from `idata`, rather than having it specified in `var_names`?

**Reply by zaxtax** [reply]
Yes! Exactly!

**Reply by zaxtax** [reply]
I should clarify that I want to keep the `var_names` override, but if the kwarg isn't given it would be nice if the `idata` was inspected to determine what was observed.

**Reply by twiecki** [reply]
CC @ricardoV94

**Reply by ricardoV94** [reply]
I'm not against it

**Comment by zaxtax** [comment]
Functionality was merged in https://github.com/pymc-devs/pymc/pull/7641
